extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/tuple/ArrayOfDoublesSketchBuildBufferAggregator.java

```diff
@@ -0,0 +1,150 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.aggregation.datasketches.tuple;
+
+import com.google.common.util.concurrent.Striped;
+import com.yahoo.memory.WritableMemory;
+import com.yahoo.sketches.tuple.ArrayOfDoublesSketches;
+import com.yahoo.sketches.tuple.ArrayOfDoublesUpdatableSketch;
+import com.yahoo.sketches.tuple.ArrayOfDoublesUpdatableSketchBuilder;
+import io.druid.query.aggregation.BufferAggregator;
+import io.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import io.druid.segment.BaseDoubleColumnValueSelector;
+import io.druid.segment.DimensionSelector;
+import io.druid.segment.data.IndexedInts;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+
+public class ArrayOfDoublesSketchBuildBufferAggregator implements BufferAggregator
+{
+
+  private static final int NUM_STRIPES = 64; // for locking per buffer position
+
+  private final DimensionSelector keySelector;
+  private final BaseDoubleColumnValueSelector[] valueSelectors;
+  private final int nominalEntries;
+  private final int maxIntermediateSize;
+  private final double[] values; // for sketch update call
+  private Striped<ReadWriteLock> stripedLock = Striped.readWriteLock(NUM_STRIPES);
+
+  public ArrayOfDoublesSketchBuildBufferAggregator(
+      final DimensionSelector keySelector,
+      final List<BaseDoubleColumnValueSelector> valueSelectors,
+      int nominalEntries,
+      int maxIntermediateSize
+  )
+  {
+    this.keySelector = keySelector;
+    this.valueSelectors = valueSelectors.toArray(new BaseDoubleColumnValueSelector[valueSelectors.size()]);
+    this.nominalEntries = nominalEntries;
+    this.maxIntermediateSize = maxIntermediateSize;
+    values = new double[valueSelectors.size()];
+  }
+
+  @Override
+  public void init(final ByteBuffer buf, final int position)
+  {
+    final WritableMemory mem = WritableMemory.wrap(buf);
+    final WritableMemory region = mem.writableRegion(position, maxIntermediateSize);
+    new ArrayOfDoublesUpdatableSketchBuilder().setNominalEntries(nominalEntries)
+        .setNumberOfValues(valueSelectors.length)
+        .setNumberOfValues(valueSelectors.length).build(region);
+  }
+
+  /**
+   * This method uses locks because Druid can call aggregate() and get() concurrently
+   * https://github.com/druid-io/druid/pull/3956
+   */
+  @Override
+  public void aggregate(final ByteBuffer buf, final int position)
+  {
+    for (int i = 0; i < valueSelectors.length; i++) {
+      values[i] = valueSelectors[i].getDouble();
+    }
+    final IndexedInts keys = keySelector.getRow();
+    final WritableMemory mem = WritableMemory.wrap(buf);
+    final WritableMemory region = mem.writableRegion(position, maxIntermediateSize);
+    final Lock lock = stripedLock.get(Objects.hash(buf, position)).writeLock();
```

Comment:
`Objects.hash(buf, ...)` calls to ByteBuffer.hashCode(), that hashes the buffer contents. It's not what we are looking for.

However, it seems that it could be safely assumed in all situations that Whatever `ByteBuffer` is passed to `init()` method, only the same ByteBuffer instance is passed later to `aggregate()` and `get()`. @gianm, @akashdw - could you please verify? Except for `relocate()`, but it's called only in group by engine, where those locks are irrelevant anyway.

So you could call here just
```java
stripedLock.getAt(lockIndex(position))
...

int lockIndex(long position)
{
  return Long.hashCode(position * LONG_PHI_MAGIC) % NUM_STRIPES;
}
```

Where LONG_PHI_MAGIC is defined as [here](https://github.com/OpenHFT/SmoothieMap/blob/80a6a2f9dcba246cdf6c6c01259c7fca90a2fd68/src/main/java/net/openhft/smoothie/SmoothieMap.java#L395).

Reply:
ok

Reply:
On a second thought, I agree that buffer is not needed for the stripe computation. Even if buffer changes, it is still fine to have the same lock. However, I don't see any need for that magic thing, which seems to help distributing lower bits to high bits. We are going to use just a few lower bits anyway. By the way, modulus can be replaced by masking if we stick with powers of 2.

Reply:
Without position mixing, you will use just `1 / GCD(sketch size(s), 64)` stripes instead of 64.

> By the way, modulus can be replaced by masking if we stick with powers of 2.

Hotspot JIT does this automatically when divisor is a compile-time constant and a power of two.

Reply:
How about keeping it simple and letting the lock do its own magic?:
stripedLock.get(position)

Reply:
Let's avoid boxing. You could apply ["smear" hash code](https://github.com/google/guava/blob/master/guava/src/com/google/common/util/concurrent/Striped.java#L536-L548) instead of "phi magic" one though, if you feel more confident about it.

Reply:
ok